File: scripts/dev-clear.ts
/**
 * Wipe all campaign-related data from the dev database.
 *
 * Usage:
 *   npx tsx scripts/dev-clear.ts            # interactive confirmation
 *   npx tsx scripts/dev-clear.ts --force     # skip confirmation
 *
 * Shortcut:
 *   npm run dev:clear
 *   npm run dev:clear -- --force
 *
 * Safety: refuses to run if NODE_ENV is "production" or if the MONGODB_URI
 * contains "prod" anywhere in the string.
 */
import mongoose from 'mongoose';
import * as readline from 'node:readline';
import * as fs from 'node:fs';
import * as path from 'node:path';

// ---------------------------------------------------------------------------
// Safety guard
// ---------------------------------------------------------------------------
function assertNotProduction(uri: string): void {
  if (process.env.NODE_ENV === 'production') {
    console.error('Refusing to run in production.');
    process.exit(1);
  }
  if (/prod/i.test(uri)) {
    console.error('MONGODB_URI looks like a production connection string. Aborting.');
    process.exit(1);
  }
}

// ---------------------------------------------------------------------------
// Confirmation prompt
// ---------------------------------------------------------------------------
async function confirm(message: string): Promise<boolean> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => {
    rl.question(`${message} (y/N) `, (answer) => {
      rl.close();
      resolve(answer.trim().toLowerCase() === 'y');
    });
  });
}

// ---------------------------------------------------------------------------
// Collections to wipe — everything tied to campaigns
// ---------------------------------------------------------------------------
const COLLECTIONS_TO_CLEAR = [
  'campaigns',
  'sessions',
  'characters',
  'players',
  'messages',
  'dicerolls',
  'notes',
  'races',
  'rules',
  'gmscreens',
  'tags',
] as const;

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main(): Promise<void> {
  const uri = process.env.MONGODB_URI;
  if (!uri) {
    console.error('MONGODB_URI is not set.');
    process.exit(1);
  }

  assertNotProduction(uri);

  const force = process.argv.includes('--force');

  if (!force) {
    console.log('\nThis will DELETE all data from the following collections:');
    COLLECTIONS_TO_CLEAR.forEach((c) => console.log(`  - ${c}`));
    console.log(`\nTarget: ${uri.replace(/\/\/[^@]+@/, '//<credentials>@')}\n`);

    const ok = await confirm('Proceed?');
    if (!ok) {
      console.log('Aborted.');
      return;
    }
  }

  await mongoose.connect(uri, { autoIndex: false });

  const db = mongoose.connection.db;
  if (!db) {
    console.error('Failed to get database handle.');
    process.exit(1);
  }

  const existing = (await db.listCollections().toArray()).map((c) => c.name);

  let totalDeleted = 0;
  for (const name of COLLECTIONS_TO_CLEAR) {
    if (!existing.includes(name)) {
      console.log(`  skip  ${name} (does not exist)`);
      continue;
    }
    const result = await db.collection(name).deleteMany({});
    totalDeleted += result.deletedCount;
    console.log(`  clear ${name} — ${result.deletedCount} documents removed`);
  }

  // Also clear campaign references from users (but don't delete users)
  const userResult = await db.collection('users').updateMany({}, { $set: { campaigns: [] } });
  console.log(`  patch users — cleared campaign refs from ${userResult.modifiedCount} user(s)`);

  // Clean up seed images from public/uploads/campaigns/
  const uploadsDir = path.resolve(process.cwd(), 'public', 'uploads', 'campaigns');
  if (fs.existsSync(uploadsDir)) {
    const files = fs.readdirSync(uploadsDir).filter((f) => f.startsWith('seed-'));
    for (const file of files) {
      fs.unlinkSync(path.join(uploadsDir, file));
    }
    if (files.length > 0) {
      console.log(`  clean ${files.length} seed image(s) from public/uploads/campaigns/`);
    }
  }

  console.log(
    `\nDone. ${totalDeleted} documents deleted across ${COLLECTIONS_TO_CLEAR.length} collections.`
  );

  await mongoose.disconnect();
}

main().catch((err) => {
  console.error('Fatal:', err);
  process.exitCode = 1;
});

File: scripts/dev_clear.py
#!/usr/bin/env python3
"""
Reset the dev environment to a clean slate.

Wipes EVERYTHING tied to the previous environment so tests start fresh:
  - every MongoDB collection — chat messages, dice rolls, campaigns, sessions,
    characters, monsters, maps, tokens, GM screens, notes, and anything else.
    User accounts are PRESERVED because they live in the graph, which this script
    does not touch: the seed needs them to exist and you need to stay logged in.
  - all locally-served upload files under public/uploads/.
  - all objects in the R2 (S3) object store.

Enumerating live collections (rather than a hardcoded list) means new
collections are wiped automatically and name drift can't leave data behind.

Usage:
    scripts/.venv/bin/python scripts/dev_clear.py            # interactive confirmation
    scripts/.venv/bin/python scripts/dev_clear.py --force    # skip confirmation

Shortcut:
    npm run dev:clear
    npm run dev:clear -- --force

Safety: refuses to run if NODE_ENV is "production", if MONGODB_URI contains
"prod", or if R2_BUCKET contains "prod".
"""

import os
import re
import shutil
import sys
from pathlib import Path

from dotenv import load_dotenv
from pymongo import MongoClient
from pymongo.errors import ConfigurationError

from r2_util import get_r2_client, r2_env

load_dotenv()

# Repo root anchored to this script's location (scripts/ is one level down)
REPO_ROOT = Path(__file__).resolve().parent.parent

# Collections we never wipe. None any more: accounts moved to the graph, which this
# script does not touch, so they survive a reset and you stay logged in. A `users`
# collection left over from before is stale placeholder data and is cleared.
PRESERVE_COLLECTIONS: set[str] = set()

# Collections no application model uses any more. Emptying would leave a shell that
# looks like live data, so they are dropped.
RETIRED_COLLECTIONS = {"users"}


# ---------------------------------------------------------------------------
# Safety
# ---------------------------------------------------------------------------

def require_mongo_uri() -> str:
    if os.environ.get("NODE_ENV") == "production":
        sys.exit("Refusing to run in production.")
    uri = os.environ.get("MONGODB_URI")
    if not uri:
        sys.exit("MONGODB_URI is not set.")
    if re.search(r"prod", uri, re.IGNORECASE):
        sys.exit("MONGODB_URI looks like a production connection string. Aborting.")
    return uri


def get_db(uri: str):
    client: MongoClient = MongoClient(uri)
    db_name = os.environ.get("MONGODB_DB")
    if db_name:
        return client, client[db_name]
    try:
        return client, client.get_default_database()
    except ConfigurationError:
        sys.exit(
            "MONGODB_URI does not include a database name and MONGODB_DB is not set.\n"
            "Either add a database name to the URI (e.g. mongodb+srv://…/cartyx) "
            "or set MONGODB_DB=cartyx in your .env file."
        )


# ---------------------------------------------------------------------------
# Clear steps
# ---------------------------------------------------------------------------

def clear_database(db) -> int:
    """Delete every document from every collection except preserved ones."""
    total = 0
    for name in sorted(db.list_collection_names()):
        if name in PRESERVE_COLLECTIONS or name.startswith("system."):
            print(f"  keep  {name}")
            continue
        if name in RETIRED_COLLECTIONS:
            db.drop_collection(name)
            print(f"  drop  {name} — retired; its data lives in the graph now")
            continue
        result = db[name].delete_many({})
        total += result.deleted_count
        print(f"  clear {name} — {result.deleted_count} documents removed")
    return total


def clear_local_uploads() -> int:
    """Remove every locally-served upload under public/uploads/ (keep the dir)."""
    uploads_dir = REPO_ROOT / "public" / "uploads"
    if not uploads_dir.is_dir():
        print("  skip  public/uploads/ (does not exist)")
        return 0
    removed = 0
    for entry in uploads_dir.iterdir():
        if entry.is_dir():
            shutil.rmtree(entry)
        else:
            entry.unlink()
        removed += 1
    print(f"  clear public/uploads/ — {removed} entr(ies) removed")
    return removed


def clear_r2_bucket() -> int:
    """Delete every object in the R2 bucket. No-op if R2 isn't configured."""
    env = r2_env()  # shared env detection + prod-bucket guard (r2_util)
    if not env:
        print("  skip  R2 (not configured — R2_* env vars missing)")
        return 0
    bucket = env["R2_BUCKET"]

    s3 = get_r2_client(env)
    deleted = 0
    for page in s3.get_paginator("list_objects_v2").paginate(Bucket=bucket):
        objs = page.get("Contents", [])
        # delete_objects accepts up to 1000 keys per request.
        for i in range(0, len(objs), 1000):
            batch = [{"Key": o["Key"]} for o in objs[i : i + 1000]]
            if not batch:
                continue
            s3.delete_objects(Bucket=bucket, Delete={"Objects": batch, "Quiet": True})
            deleted += len(batch)
    print(f"  clear R2 bucket '{bucket}' — {deleted} object(s) removed")
    return deleted


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    uri = require_mongo_uri()
    force = "--force" in sys.argv
    bucket = os.environ.get("R2_BUCKET") or "(not configured)"

    if not force:
        print("\nThis will PERMANENTLY DELETE, for a clean test environment:")
        print("  - every MongoDB collection (accounts live in the graph and are kept)")
        print("  - all files under public/uploads/")
        print(f"  - all objects in the R2 bucket '{bucket}'")
        masked = re.sub(r"//[^@]+@", "//<credentials>@", uri)
        print(f"\nMongo target: {masked}\n")
        if input("Proceed? (y/N) ").strip().lower() != "y":
            print("Aborted.")
            return

    client, db = get_db(uri)
    try:
        print("\nDatabase:")
        total = clear_database(db)
        print("\nLocal uploads:")
        clear_local_uploads()
        print("\nR2 object store:")
        clear_r2_bucket()
        print(
            f"\nDone. {total} documents deleted; public/uploads/ and the R2 bucket emptied. "
            "User accounts preserved."
        )
    finally:
        client.close()


if __name__ == "__main__":
    main()
